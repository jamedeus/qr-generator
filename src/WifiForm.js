import React from 'react';
import PropTypes from 'prop-types';
import Form from 'react-bootstrap/Form';
import FloatingLabel from 'react-bootstrap/FloatingLabel';
import Button from 'react-bootstrap/Button';

function WifiForm({ generate, validated }) {
    return (
        <Form noValidate validated={validated} onSubmit={generate}>
            <FloatingLabel label="SSID" className="mb-3">
                <Form.Control
                    type="text"
                    name="ssid"
                    placeholder="SSID"
                    required
                />
            </FloatingLabel>

            <FloatingLabel label="Password" className="mb-3">
                <Form.Control
                    type="password"
                    name="password"
                    placeholder="Password"
                    required
                />
            </FloatingLabel>

            <div className="d-flex">
                <Button variant="primary" type="submit" className="mx-auto">
                    Generate
                </Button>
            </div>
        </Form>
    );
}

WifiForm.propTypes = {
    generate: PropTypes.func,
    validated: PropTypes.bool
};

export default WifiForm;
