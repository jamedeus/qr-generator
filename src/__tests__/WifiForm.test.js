import renderer from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react';
import WifiForm from '../WifiForm';

describe('WifiForm', () => {
    // Mock generate hook
    const mockGenerate = jest.fn(e => e.preventDefault());

    it('matches snapshot', () => {
        const component = renderer.create(
            <WifiForm generate={mockGenerate} validated={false} />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls generate function when button clicked', () => {
        const { getByText } = render(
            <WifiForm generate={mockGenerate} validated={false} />
        );

        const generateButton = getByText(/generate/i);
        fireEvent.click(generateButton);

        expect(mockGenerate).toHaveBeenCalled();
    });

    it('shows validation highlights when validated is true', () => {
        const { getByText } = render(
            <WifiForm generate={mockGenerate} validated={true} />
        );

        // Get form element, confirm correct type
        const form = getByText(/generate/i).parentElement.parentElement;
        expect(form.tagName).toBe('FORM');

        expect(form.classList).toContainEqual('was-validated');
    });

    it('does not show validation highlights when validated is false', () => {
        const { getByText } = render(
            <WifiForm generate={mockGenerate} validated={false} />
        );

        // Get form element, confirm correct type
        const form = getByText(/generate/i).parentElement.parentElement;
        expect(form.tagName).toBe('FORM');

        expect(form.classList).not.toContainEqual('was-validated');
    });

    it('requires all fields', async () => {
        const { getByPlaceholderText } = render(
            <WifiForm generate={mockGenerate} validated={false} />
        );

        // Get reference to each field
        const ssidField = getByPlaceholderText('SSID');
        const passField = getByPlaceholderText('Password');

        // All empty fields should be invalid
        expect(ssidField.validity.valueMissing).toBeTruthy();
        expect(passField.validity.valueMissing).toBeTruthy();

        // Add arbitrary text, should be valid
        ssidField.value = "foobar";
        expect(ssidField.validity.valid).toBeTruthy();
        passField.value = "foobar";
        expect(passField.validity.valid).toBeTruthy();
    });
});
