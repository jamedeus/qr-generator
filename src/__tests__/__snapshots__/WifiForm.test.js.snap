// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WifiForm matches snapshot 1`] = `
<form
  className=""
  noValidate={true}
  onSubmit={[MockFunction]}
>
  <div
    className="mb-3 form-floating"
  >
    <input
      className="form-control"
      name="ssid"
      placeholder="SSID"
      required={true}
      type="text"
    />
    <label>
      SSID
    </label>
  </div>
  <div
    className="mb-3 form-floating"
  >
    <input
      className="form-control"
      name="password"
      placeholder="Password"
      required={true}
      type="password"
    />
    <label>
      Password
    </label>
  </div>
  <div
    className="d-flex"
  >
    <button
      className="mx-auto btn btn-primary"
      disabled={false}
      type="submit"
    >
      Generate
    </button>
  </div>
</form>
`;
